modal/functions.py

```diff
@@ -176,6 +178,45 @@ async def _create_input(args, kwargs, client, idx=None) -> api_pb2.FunctionPutIn
         )
 
 
+async def _stream_function_call_data(
+    client, function_call_id: str, variant: Literal["data_in", "data_out"]
+) -> AsyncIterator[Any]:
+    """Read from the `data_in` or `data_out` stream of a function call."""
+    last_index = 0
+    retries_remaining = 10
+
+    if variant == "data_in":
+        stub_fn = client.stub.FunctionCallGetDataIn
+    elif variant == "data_out":
+        stub_fn = client.stub.FunctionCallGetDataOut
+    else:
+        raise ValueError(f"Invalid variant {variant}")
+
+    while True:
+        req = api_pb2.FunctionCallGetDataRequest(function_call_id=function_call_id, last_index=last_index)
+        try:
+            async for chunk in unary_stream(stub_fn, req):
+                if chunk.index <= last_index:
+                    continue
+                last_index = chunk.index
+                if chunk.data_blob_id:
+                    message_bytes = await blob_download(chunk.data_blob_id, client.stub)
+                else:
+                    message_bytes = chunk.data
+                message = deserialize_data_format(message_bytes, chunk.data_format, client)
+                yield message
+        except (GRPCError, StreamTerminatedError) as exc:
+            if retries_remaining > 0:
```

Comment:
Should we reset `retries_remaining` if there was no exception? (in `else`?) Otherwise a really long-running generator could naturally exhaust the 10 retries, even if they're spread apart.

Reply:
It seems reasonable, but I also think limiting the number of retries is safer for load shedding during periods of high traffic.

If we aim to have (1-eps) reliability for this RPC, then 10 retries gives us (1/eps)^10 reliability

Reply:
Yeah that's a good point. It's just that we have a natural failure rate per individual RPC. Since longer-lived generators will require this function to do a few orders of magnitudes more RPC calls, longer-lived generators will have a higher overall failure rate, which feels weird.

Reply:
There are only 1440 minutes in a day!

Reply:
Fair enough, let's leave it as is